### PR TITLE
webgpu: add backward kernels

### PR DESCRIPTION
**Pull request by ngxson**
## Overview

A fun experiment to see if we can finetune a model 100% on browser without install anything locally

Requires https://github.com/ggml-org/llama.cpp/pull/27199 , some changes in this PR will be removed once that PR is merged

CC @ggml-org/ggml-webgpu if you think this is something acceptable, or just adding too much maintenance burdens

Demo: https://huggingface.co/spaces/ngxson/wllama-finetune-demo

Benchmark so far:

model | ctx | work | CUDA | WebGPU | ratio (slower by)
-- | -- | -- | -- | -- | --
stories260K-f32 | 128 | 20 epochs | 3.45 s | 11.31 s | 3.3x
Qwen3-0.6B-f32 | 256 | 1 epoch | 9.65 s | 62.62 s | 6.5x

<br class="Apple-interchange-newline">

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: 100% AI-generated

### CODE REVIEW

**Comment by yomaytk** [comment]
Finetuning the model in the browser is interesting, and supporting the ops for this use case is fine with me.

The changes look good to me. That said, it seems the performance of some ops (e.g. `REPEAT_BACK`, `OUT_PROD`) can be improved, so once this PR is merged I'll follow up in a separate PR when I have a chance. Does that sound good?

**Comment by ngxson** [comment]
@yomaytk thanks, yes I agree: any improvements can be in follow-up PR. this is meant to be an initial demo one to see if the shape is good.

btw, I finally managed to put together a simple demo here: https://huggingface.co/spaces/ngxson/wllama-finetune-demo

I'm using the super small, super dumb model for demo purpose, but we should get something more useful once I implement the LoRA training mode, will be super exciting

File: ggml/src/ggml-webgpu/ggml-webgpu.cpp [hunk 1]
```diff
@@ -4625,6 +4866,34 @@ static bool ggml_backend_webgpu_device_supports_op(ggml_backend_dev_t dev, const
             supports_op = (op->type == GGML_TYPE_F32 || op->type == GGML_TYPE_F16) &&
                           (src0->type == GGML_TYPE_F32 || src0->type == GGML_TYPE_F16);
             break;
+        case GGML_OP_OPT_STEP_ADAMW:
+        case GGML_OP_OPT_STEP_SGD:
+            supports_op = src0->type == GGML_TYPE_F32 && ggml_is_contiguous(src0);
```
Comment (yomaytk) [inline ggml/src/ggml-webgpu/ggml-webgpu.cpp:4871]:
```suggestion
            {
                const int n_src = op->op == GGML_OP_OPT_STEP_ADAMW ? 5 : 3;
                supports_op = true;
                for (int i = 0; i < n_src; ++i) {
                    if (op->src[i]->type != GGML_TYPE_F32 || !ggml_is_contiguous(op->src[i])) {
                        supports_op = false;
                        break;
                    }
                }
            }
```

I'm not sure any real workload hits the path with non-f32 or non-contiguous tensors here, but the shader assumes all of them are f32 and contiguous, so I think it is better to check every src rather than only `src0`.

File: ggml/src/ggml-webgpu/ggml-webgpu.cpp [hunk 2]
```diff
@@ -4625,6 +4866,34 @@ static bool ggml_backend_webgpu_device_supports_op(ggml_backend_dev_t dev, const
...
+        case GGML_OP_RMS_NORM_BACK:
+            supports_op = op->type == GGML_TYPE_F32 && src0->type == GGML_TYPE_F32 && src1->type == GGML_TYPE_F32 &&
+                          ggml_is_contiguous(src0) && ggml_is_contiguous(src1) && ggml_is_contiguous(op);
+            break;
+        case GGML_OP_REPEAT_BACK:
+            supports_op = op->type == GGML_TYPE_F32 && src0->type == GGML_TYPE_F32 && ggml_is_contiguous(op);
+            break;
+        case GGML_OP_GET_ROWS_BACK:
+            supports_op = op->type == GGML_TYPE_F32 && src0->type == GGML_TYPE_F32 && src1->type == GGML_TYPE_I32 &&
+                          ggml_is_contiguous(op);
```
Comment (yomaytk) [inline ggml/src/ggml-webgpu/ggml-webgpu.cpp:4895]:
```suggestion
                          ggml_is_contiguous_rows(src0) && ggml_is_contiguous(op);
```
Adding this check would be better since the shader assumes that each row of `src0` is contiguous.

**Comment by yomaytk** [comment]
I suggested two changes, could you take a look at them?

Also, the demo is nice :) I'll give it a try.

File: ggml/src/ggml-webgpu/ggml-webgpu.cpp (hunk 2 above)
Comment (ngxson) [reply ggml/src/ggml-webgpu/ggml-webgpu.cpp:4895]: actually we don't need get_rows_back anymore, I'll revert this
